### PR TITLE
 Disallow Ed448 signature maleability

### LINKED ISSUE #7706: ed448 signature malleability
Similar to #7693, ed448 verification doesn't check that `s` is less than the group order, allowing malleability for some signatures. Wycheproof doesn't have vectors for this yet, but I generated an example:

```c
#include <assert.h>
#include <stdio.h>
#include <string.h>
#include <openssl/crypto.h>
#include <openssl/evp.h>
#include <openssl/x509.h>


int main(void) {
    OPENSSL_init();
    int ret = 0;
    int res;
    unsigned char *public = OPENSSL_hexstr2buf("fedb02a658d74990244d9d10cf338e977565cbbda6b24c716829ed6ee1e4f28cf2620c052db8d878f6243bffc22242816c1aaa67d2f3603600", NULL);
    unsigned char *sig = OPENSSL_hexstr2buf("0cc16ba24d69277f927c1554b0f08a2a711bbdd20b058ccc660d00ca13542a3cef9e5c44c54ab23a2eb14f947e167b990b080863e28b399380f30db6e54d5d1406d23378ffde11b1fb81b2b438a3b8e8aa7f7f4e1befcc905023fab5a5465053844f04cf0c1b51d84760f869588687f57500", NULL);
    unsigned char *msg = OPENSSL_hexstr2buf("38", NULL);

    EVP_PKEY *pkey = EVP_PKEY_new_raw_public_key(NID_ED448, NULL, public, 57);
    assert(pkey != NULL);
    EVP_MD_CTX *ctx = EVP_MD_CTX_new();
    res = EVP_DigestVerifyInit(ctx, NULL, NULL, NULL, pkey);
    assert(res == 1);
    res = EVP_DigestVerify(ctx, sig, 114, msg, 1);
    if (res == 1) {
        printf("Verification passed but should have failed.\n");
        ret = 1;
    }
    exit(ret);
}
```

### PR DESCRIPTION
**Pull request by mattcaswell**
Check that s is less than the order before attempting to verify the
signature as per RFC8032 5.2.7

Fixes #7706

### CODE REVIEW

**Comment by mattcaswell** [comment]
Ping?

File: crypto/ec/curve448/eddsa.c [hunk 1]
```diff
@@ -246,10 +246,36 @@ c448_error_t c448_ed448_verify(
...
+    int i;
+
+    /*
+     * Check that s (second 57 bytes of the sig) is less than the order. Both
+     * s and the order are in little-endian format. This can be done in
+     * variable time, since if this is not the case the signature if publicly
+     * invalid.
+     */
+    for (i = EDDSA_448_PUBLIC_BYTES - 1; i >= 0; i--) {
+        if (signature[i + EDDSA_448_PUBLIC_BYTES] > order[i])
```
Comment (kroeckx) [inline crypto/ec/curve448/eddsa.c]: So you're starting at `signature[EDDSA_448_PUBLIC_BYTES - 1 + EDDSA_448_PUBLIC_BYTES]` which seems wrong. Please consider putting `order` in the same byte order as `signature`

Comment (kroeckx) [reply crypto/ec/curve448/eddsa.c]: Oh, I guess I'm just confused between EDDSA_448_PUBLIC_BYTES and EDDSA_448_SIGNATURE_BYTES

**Comment by mspncp** [comment]
Typo: malleability (in commit message)

File: crypto/ec/curve448/eddsa.c [hunk 2]
```diff
@@ -246,10 +246,36 @@ c448_error_t c448_ed448_verify(
...
     curve448_scalar_t challenge_scalar;
     curve448_scalar_t response_scalar;
+    /* Order in little endian format */
+    static const uint8_t order[] = {
+        0xF3, 0x44, 0x58, 0xAB, 0x92, 0xC2, 0x78, 0x23, 0x55, 0x8F, 0xC5, 0x8D,
+        0x72, 0xC2, 0x6C, 0x21, 0x90, 0x36, 0xD6, 0xAE, 0x49, 0xDB, 0x4E, 0xC4,
+        0xE9, 0x23, 0xCA, 0x7C, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+        0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+        0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x3F, 0x00
+    };
```
Comment (kroeckx) [inline crypto/ec/curve448/eddsa.c]: This seems to be q, not the order which is 4*q

Comment (kroeckx) [reply crypto/ec/curve448/eddsa.c]: rfc8032 has it as L and the order, so I guess it's fine.

**Comment by mattcaswell** [comment]
I fixed the typo (it also appeared in the evppkey.txt file).

@kroeckx I forgot to specify that this is also for backport to 1.1.1. Please can you reconfirm?

**Comment by kroeckx** [comment]
Yes, merge it to any branch that has ed448 code

**Comment by mattcaswell** [comment]
Pushed to master and 1.1.1. Thanks.

**Pull request closed without merging**